components/bm1397/bm1366.c

```diff
@@ -0,0 +1,738 @@
+#include "freertos/task.h"
+#include "esp_log.h"
+
+#include "serial.h"
+#include "bm1366.h"
+#include "utils.h"
+#include "crc.h"
+#include "global_state.h"
+
+#define BM1366_RST_PIN  GPIO_NUM_1
+
+
+#define TYPE_JOB 0x20
+#define TYPE_CMD 0x40
+
+#define GROUP_SINGLE 0x00
+#define GROUP_ALL 0x10
+
+#define CMD_JOB 0x01
+
+#define CMD_SETADDRESS 0x00
+#define CMD_WRITE 0x01
+#define CMD_READ 0x02
+#define CMD_INACTIVE 0x03
+
+#define RESPONSE_CMD 0x00
+#define RESPONSE_JOB 0x80
+
+
+
+#define SLEEP_TIME 20
+#define FREQ_MULT 25.0
+
+#define CLOCK_ORDER_CONTROL_0 0x80
+#define CLOCK_ORDER_CONTROL_1 0x84
+#define ORDERED_CLOCK_ENABLE 0x20
+#define CORE_REGISTER_CONTROL 0x3C
+#define PLL3_PARAMETER 0x68
+#define FAST_UART_CONFIGURATION 0x28
+#define TICKET_MASK 0x14
+#define MISC_CONTROL 0x18
+
+typedef struct __attribute__((__packed__)) {
+    uint8_t preamble[2];
+    uint32_t nonce;
+    uint8_t midstate_num;
+    uint8_t job_id;
+    uint16_t version;
+    uint8_t crc;
+} asic_result;
+
+static const char *TAG = "bm1366Module";
+
+static uint8_t asic_response_buffer[CHUNK_SIZE];
+static asic_result _asic_result;
+static task_result result;
+
+/// @brief
+/// @param ftdi
+/// @param header
+/// @param data
+/// @param len
+static void _send_BM1366(uint8_t header, uint8_t * data, uint8_t data_len, bool debug) {
+    packet_type_t packet_type = (header & TYPE_JOB) ? JOB_PACKET : CMD_PACKET;
+    uint8_t total_length = (packet_type == JOB_PACKET) ? (data_len+6) : (data_len+5);
+
+    //allocate memory for buffer
+    unsigned char *buf = malloc(total_length);
+
+    //add the preamble
+    buf[0] = 0x55;
+    buf[1] = 0xAA;
+
+    //add the header field
+    buf[2] = header;
+
+    //add the length field
+    buf[3] = (packet_type == JOB_PACKET) ? (data_len+4) : (data_len+3);
+
+    //add the data
+    memcpy(buf+4, data, data_len);
+
+    //add the correct crc type
+    if (packet_type == JOB_PACKET) {
+        uint16_t crc16_total = crc16_false(buf+2, data_len+2);
+        buf[4+data_len] = (crc16_total >> 8) & 0xFF;
+        buf[5+data_len] = crc16_total & 0xFF;
+    } else {
+        buf[4+data_len] = crc5(buf+2, data_len+2);
+    }
+
+    //send serial data
+    SERIAL_send(buf, total_length, debug);
+
+    free(buf);
+}
+
+static void _send_simple(uint8_t * data, uint8_t total_length){
+    unsigned char *buf = malloc(total_length);
+    memcpy(buf, data, total_length);
+    SERIAL_send(buf, total_length, false);
+
+    free(buf);
+}
+
+static void _send_chain_inactive(void) {
+
+    unsigned char read_address[2] = {0x00, 0x00};
+    //send serial data
+    _send_BM1366((TYPE_CMD | GROUP_ALL | CMD_INACTIVE), read_address, 2, false);
+}
+
+static void _set_chip_address(uint8_t chipAddr) {
+
+    unsigned char read_address[2] = {chipAddr, 0x00};
+    //send serial data
+    _send_BM1366((TYPE_CMD | GROUP_SINGLE | CMD_SETADDRESS), read_address, 2, false);
+}
+
+static unsigned char _reverse_bits(unsigned char num) {
```

Comment:
look like a lot of helper function can be factorized, like this one. Can you created a common.c/.h in the bm1397 component ?